morango/syncsession.py

```diff
@@ -21,21 +33,157 @@ class Connection(object):
+            "instance": json.dumps(InstanceIDSerializer(InstanceIDModel.get_or_create_current_instance()[0]).data),
+            "nonce": nonce,
+        }
+
+        # sign the nonce/ID combo to attach to the request
+        message = "{nonce}:{id}".format(**data)
+        data["signature"] = client_cert.sign(message)
+
+        # Sync Session creation request
+        session_resp = self._request(api_urls.SYNCSESSION, method="POST", data=data)
+
+        # check that the nonce/id were properly signed by the server cert
+        if not server_cert.verify(message, session_resp.data.pop("signature")):
+            raise CertificateSignatureInvalid()
+
+        # if sync session successfully created, create on our own side
+        if session_resp.status_code == status.HTTP_201_CREATED:
+
+            # build the data to be used for creating our own syncsession
+            data = {
+                "id": data['id'],
+                "start_timestamp": timezone.now(),
+                "last_activity_timestamp": timezone.now(),
+                "active": True,
+                "is_server": False,
+                "local_certificate": client_cert,
+                "remote_certificate": server_cert,
+                "profile": client_cert.profile,
+                "connection_kind": "network",
+                "connection_path": self.base_url,
+                "local_ip": socket.gethostbyname(socket.gethostname()),
+                "remote_ip": socket.gethostbyname(self.base_url),
+                "local_instance": json.dumps(InstanceIDSerializer(InstanceIDModel.get_or_create_current_instance()[0]).data),
+                "remote_instance": session_resp.data.get("instance") or "{}",
+            }
+            self.sync_session = SyncSession.objects.create(**data)
+
+            return SyncClient(self)
+
+    def get_remote_certificates(self, primary_partition):
+        remote_certs = []
+        # request certs for this primary partition, where the server also has a private key for
+        remote_certs_resp = self._request(api_urls.CERTIFICATE, params={'primary_partition': primary_partition})
+
+        # inflate remote certs into a list of unsaved models
+        for cert in json.loads(remote_certs_resp.data):
+            remote_certs.append(Certificate.deserialize(cert["serialized"], cert["signature"]))
+        return remote_certs
+
+    def certificate_signing_request(self, parent_cert, scope_definition_id, scope_params, username=None, password=None):
+        csr_key = Key()
+        # build up data for csr
+        data = {
+            "parent": parent_cert.id,
+            "profile": self.profile,
+            "scope_definition": scope_definition_id,
+            "scope_version": parent_cert.scope_version,
+            "scope_params": scope_params,
+            "public_key": csr_key.get_public_key_string()
+        }
+        csr_resp = self._request(api_urls.CERTIFICATE, method="POST", data=data, auth=(username, password))
```

Comment:
The Kolibri basic auth stuff requires that you specify a facility, also -- I guess we can leave it up to the calling code to construct the multi-param username, as that's not general-purpose.

Reply:
(can discuss -- may be useful to handle some of it here, with some helpers, taking in `auth` as a `dict`)